packages/smooth_app/lib/pages/prices/generic_infinite_scroll.dart

```diff
@@ -0,0 +1,314 @@
+// generic_infinite_scroll.dart
+
+import 'package:flutter/material.dart';
+
+/// A generic class for handling infinite scrolling in lists.
+/// [T] is the type of items being displayed.
+/// [P] is the type of parameters used for pagination.
+class InfiniteScrollController<T, P> {
+  InfiniteScrollController({
+    required this.fetchItems,
+    Iterable<T> initialItems = _kEmptyIterable,
+    this.initialPage = 1,
+    this.onError,
+  })  : _currentPage = initialPage,
+        _items = List<T>.from(initialItems),
+        initialItems = List<T>.from(initialItems);
+
+  static const Iterable<Never> _kEmptyIterable = <Never>[];
+
+  /// Returns a Future with the fetched items and a boolean indicating if more items can be loaded
+  final Future<(List<T>, bool)> Function(P parameters, int page) fetchItems;
+
+  /// Parameters for the fetch operation
+  P? parameters;
+
+  /// Initial page number
+  final int initialPage;
+
+  /// Initial items to populate the list
+  final List<T> initialItems;
+
+  /// Called when an error occurs during fetching
+  final Function(dynamic error)? onError;
+
+  /// Current items in the list
+  List<T> _items;
+  List<T> get items => _items;
+
+  /// Current page being fetched
+  int _currentPage;
+  int get currentPage => _currentPage;
+
+  /// Whether currently loading more items
+  bool _isLoading = false;
+  bool get isLoading => _isLoading;
+
+  /// Whether more items can be loaded
+  bool _hasMoreItems = true;
+  bool get hasMoreItems => _hasMoreItems;
```

Comment:
I'm a bit surprised that you have to maintain a `bool` field for `hasMoreItems`, as it could be computed with `_currentPage` and `totalPages`;